### PR TITLE
Fix edge cases of if/when hl. Fix generic proc hl.

### PR DESCRIPTION
**Pull request by ozra**
Looked around some nim files and couldn't see anything weird happening. Ping me if you find something and I can be of help.

### CODE REVIEW

**Comment by ozra** [comment]
Now it seems pretty solid.
Checked with standard monokai color theme to ensure it looks as before, but fixed.
Along with additional possible highlightings on the more detailed scopes.

**Comment by ozra** [comment]
The .nim suffixes shouldn't be necessary because of nesting syntax scope, but they're left in now as "safety", will examine if they can be safely removed at later refinments.

**Comment by onionhammer** [comment]
There's still some issues here; 

```nim
## Pushbullet API for Nim
when not defined(ssl):
    {.define: ssl.}
```

'when' and 'not' are both keywords, but aren't being highlighted; 'when' should be highlighted exactly the same as 'if'

**Comment by ozra** [comment]
Just discovered this minutes ago, the RE assumed leading indent for some control constructs. I'll update the PR soon.

**Comment by dom96** [comment]
Please make sure it works well. Github uses this grammar now: https://github.com/github/linguist/tree/master/vendor/grammars

**Comment by onionhammer** [comment]
Oh, well in *that* case lol..

**Comment by dom96** [comment]
Seriously guys. Look at this shit:

capture

**Comment by Varriount** [comment]
Wait, what? So NimLime *is* being used for Github's syntax highlighting?

**Comment by dom96** [comment]
Yes!

**Comment by Varriount** [comment]
I should probably tell @ozra that, since I specifically in another issue that NimLime isn't connected with GitHub highlighting...

**Comment by ozra** [comment]
Okay. Then we got that sorted out. I'll get on it asap. 
The grammars should probably be forked out to its own repo; Github and sublime are too different to target to get more detail in (which you may very well want in the editor, but very likely not on github). Would anyone care to "hardfork" the repo to `nim-lang` and call it github-highlighter or something? It could keep the basic nimlime related stuff just to ease development. I'd love to help out with the grammars, and getting what crap I introduced out, but I'm still a bit fumbly with githubbing.

It would be easier to fix hl for github first, then add fixes to nimlime, cause there are scopes there that github seems to dislike but that I rely on in sublime...

So right now, I'll just work on a new local repo, await a new github repo, clone that and copy the adjusted grammar into it and PR. Ok?

**Comment by dom96** [comment]
I asked Github to roll it back: https://github.com/github/linguist/issues/2468

**Comment by pchaigno** [comment]
> Github and sublime are too different to target to get more detail in

They're using different regular expression engines but that should only impact really specific cases.
<br>

> cause there are scopes there that github seems to dislike but that I rely on in sublime...

GitHub is following [naming conventions](https://manual.macromates.com/en/language_grammars) normally. If it's not I think you should report it ;)
<br>

@dom96 @Varriount Is there an issue opened for the bug @dom96 reported? Would it be quick to fix? If it is, I'm confident they'd be okay to push a quick update at [github/linguist](https//github.com/github/linguist). If not we can roll it back as suggested :/

**Comment by ozra** [comment]
Ok, I'll do like this simply: I'll update directly to NimLime, and then things can go from there, and meanwhile, the old, but "less github-buggy", version is hopefully quickly rolled back by github then.

**Comment by ozra** [comment]
Reduced and fixes. There might still be something, but I've looked through "heavy" files in the compiler to see if I spot goofy behaviour and haven't stumbled upon any more (number literals, procs, conditionals, etc. seems good). And all additional scope has been removed in favour of github compat.

**Comment by pchaigno** [comment]
@ozra So this pull request fixes the issue reported by @dom96?

**Comment by ozra** [comment]
@pchaigno - it should. The previous additions added nested scopes, which works fine in sublime, allowing the user to refine coloring at will, but github reacted to that in a strange way coloring some things and not others.

I removed them all except for the possibility to color "primary" and "secondary" keywords differently (which is done without nesting and should not cause problems for github).

**Comment by pchaigno** [comment]
@ozra Awesome!

@Varriount Do you think this could be reviewed quickly? Once it's merged I can open a pull request at github/linguist to update the grammar. I'll ping the GitHubbers and we could probably have an update deployed in the day following the merge :)

**Comment by dom96** [comment]
Please make sure that it works in diffs too!

**Comment by ozra** [comment]
@pchaigno cool.

@dom96 I have no idea how to test github highlighting without actually going live with it :-/ Simply testing the showcased diff bug in sublime won't do - because there's nothing wrong with the text in that diff, so that's a peculiar one. Hopefully, it just had to do with the nested scopes..

**Comment by pchaigno** [comment]
> I have no idea how to test github highlighting without actually going live with it 

@dom96 You can use Lightshow for that: https://github-lightshow.herokuapp.com/

**Comment by dom96** [comment]
@ozra Actually I think the issue I am seeing is incorrect highlighting of ``0'i16`` literals.

**Comment by ozra** [comment]
Hmmm. might be something like that. if I type `0'i16` in sublime it looks ok, if I add, say `a0'i16` or something else non syntactical - the area turns red (well bloody pink). So if differ takes a num literal out of context (cut up or so) and highlights, it will be seen as an un terminated string and display "error color". Might be possible to make the matching "safer"..

**Comment by ozra** [comment]
I made the number literal suffix matcher veeery lenient, requiring only enough literal to establish that it _most likely_ is a number literal, ignoring syntactic correctness. This way, even if it's chopped off, missing information, it is more likely that it still is rendered as a num literal. It doesn't affect any other correct code, so seems like a safe bet.

File: Syntaxes/Nim.YAML-tmLanguage [hunk 1]
```diff
@@ -60,17 +60,15 @@ patterns:
...
   name: meta.proc.nim
   patterns:
-  - begin: (?<![\w\x{80}-\x{10FFFF}])(proc|method|template|macro|iterator|converter)\s+(\`([^\s\`]+)\`|([\w\x{80}-\x{10FFFF}]+))(\*)?(\(.*\))?(\:[^={]+)?
+  - begin: (?<![\w\x{80}-\x{10FFFF}])(proc|method|template|macro|iterator|converter)\s+((\`[^\s\`]+\`)|([\w\x{80}-\x{10FFFF}]+))(\*)?
     end: (\=)
     beginCaptures:
       '1': {name: storage.type.proc.nim}
       '2': {name: entity.name.function.nim}
-      '5': {name: keyword.secondary.operator.kw-public.nim}
-      '6': {name: variable.parameter.nim}
```
Comment (Varriount) [inline Syntaxes/Nim.YAML-tmLanguage]: Why remove these?

File: Syntaxes/Nim.YAML-tmLanguage [hunk 2]
```diff
@@ -116,80 +115,67 @@ patterns:
 
 - comment: A float literal
   name: constant.numeric.float.decimal.nim
-  match: (?<![\w\x{80}-\x{10FFFF}])(\d[_\d]*((\.[_\d]+([eE][\+\-]?\d[_\d]*)?)|([eE][\+\-]?\d[_\d]*)))('?([fF](32|64|128))|[fFdD])?   #' highlight-flaw-fixer-comment
```
Comment (Varriount) [inline Syntaxes/Nim.YAML-tmLanguage]: What was wrong with the `[iIuUfF](8|16|32|64)` segments for this line and the others below it?

File: Syntaxes/Nim.YAML-tmLanguage [hunk 3]
```diff
@@ -251,7 +237,7 @@ patterns:
 - comment: (Raw) Triple Quoted String
   name: string.quoted.triple.nim
   begin: r?\"\"\"
-  end: \"\"\"[^\"]   #" highlight-flaw-fixer-comment
```
Comment (Varriount) [inline Syntaxes/Nim.YAML-tmLanguage]: Does the `#" highlight-flaw-fixer-comment` mean something, or have any particular purpose here and elsewhere?

**Comment by dom96** [comment]
@ozra Here is an example of the error: https://github.com/nim-lang/Nim/pull/2976/files#diff-3a2daa5a0e408529216e7d6e0900bc29L419

**Comment by ozra** [comment]
@Varriount
- The highlight-flaw things are comments - they were there to fix sublimes YAML highlighting problems, LOL. Not meant to be part of the PR, only for use while editing - it was all monochrome without them :-/ - Didn't want to jump on Yaml highlighter fixing too ;-)
- The change on num literals was because of @dom96 reported trouble in diffs - so I made a 'safest parsing' possible approach to it. The change reduces the chances of "all pink" attack if num-literals were to be cut in the middle. We could probably revert that change though, it was in the last push.

@dom96 - Can't reproduce it locally - so it should work after the this.

**Comment by Varriount** [comment]
@ozra and what about the removal of the captures in the section after line 65?

**Comment by ozra** [comment]
@Varriount - don't know why GitHub displays two removed, only 'variable.parameter'nim' was removed, because that match in the RE was removed. It's handled by the content "patterns", with the earlier scope names, as a safety precaution to make GitHub happy.

**Comment by onionhammer** [comment]
Still finding issues here and there;

an example;

```

    # Imports
    import os, parseopt2, uri
    var x1 = 5
```

the numbers are highlighted as a number, even though they're part of an identifier.

**Comment by ozra** [comment]
I'll have a look - maybe it was that last push "to parse numbers on the safe side", which was probably completely unneeded anyway.

**Comment by ozra** [comment]
Yep. It was that. Re-introduced previous number lexing. See PR #39.

**Comment by onionhammer** [comment]
Okay; next issue;

inconsistent colon (:) highlighting for

```nim
if something:
for something:
# etc
```

the colon is highlighted as white

but for

```nim
try: discard
finally: discard
```

the colon is red

**Comment by dom96** [comment]
I hope you guys are not over complicating this syntax highlighter.

**Comment by ozra** [comment]
There was a difference in lookahead vs capture, giving them more specific scope classes in on of the cases. Changed them to all use lookahead. In PR #40.

**Comment by Varriount** [comment]
@dom96 You're talking about a system with limited logic that uses regexes to parse a very flexible programming language. 'complicated' doesn't even begin to describe it.

**Comment by onionhammer** [comment]
@ozra looks good so far

**Comment by ozra** [comment]
@onionhammer good - just shout out again if you find anything, I could take it on since I've got some miles on those RE's by now ;-) Preferably in a new issue to easier track it.

**Pull request merged**